server/sonar-server/src/main/java/org/sonar/server/plugins/ws/PendingAction.java

```diff
@@ -73,19 +80,49 @@ public void handle(Request request, Response response) throws Exception {
     JsonWriter jsonWriter = response.newJsonWriter();
 
     jsonWriter.beginObject();
-    writeInstalling(jsonWriter, compatiblePluginsByKey);
-    writeRemoving(jsonWriter, compatiblePluginsByKey);
+    writePlugins(jsonWriter, compatiblePluginsByKey);
     jsonWriter.endObject();
     jsonWriter.close();
   }
 
-  private void writeInstalling(JsonWriter json, Map<String, Plugin> compatiblePluginsByKey) {
-    Collection<PluginInfo> plugins = pluginDownloader.getDownloadedPlugins();
-    pluginWSCommons.writePluginInfoList(json, plugins, compatiblePluginsByKey, ARRAY_INSTALLING);
+  private void writePlugins(JsonWriter json, Map<String, Plugin> compatiblePluginsByKey) {
+    Collection<PluginInfo> uninstalledPlugins = installer.getUninstalledPlugins();
+    pluginWSCommons.writePluginInfoList(json, uninstalledPlugins, compatiblePluginsByKey, ARRAY_REMOVING);
+
+    Collection<PluginInfo> downloadedPlugins = pluginDownloader.getDownloadedPlugins();
+    Collection<PluginInfo> installedPlugins = installer.getPluginInfos();
+    List<String> installedPluginKeys = from(installedPlugins).transform(PluginInfoToKey.INSTANCE).toList();
+
+    Collection<PluginInfo> newPlugins = from(downloadedPlugins)
+      .filter(not(new MatchPluginKeys(installedPluginKeys)))
+      .toList();
+    pluginWSCommons.writePluginInfoList(json, newPlugins, compatiblePluginsByKey, ARRAY_INSTALLING);
+
+    Collection<PluginInfo> updatedPlugins = from(downloadedPlugins)
+      .filter(new MatchPluginKeys(installedPluginKeys))
+      .toList();
+    pluginWSCommons.writePluginInfoList(json, updatedPlugins, compatiblePluginsByKey, ARRAY_UPDATING);
+  }
+
+  private enum PluginInfoToKey implements Function<PluginInfo, String> {
+    INSTANCE;
+
+    @Override
+    public String apply(@Nonnull PluginInfo input) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Remove this "Nonnull" annotation to honor the overridden method's contract. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS2638)